File: 0074 search-a-2d-matrix/solution.go
func searchMatrix(matrix [][]int, target int) bool {
    m,n:=len(matrix),len(matrix[0])
    l,r:=0,m*n-1
    for l<=r {
        mid:=l+(r-l)/2
        row,col:=mid/n,mid%n
        val:=matrix[row][col]
        if val==target {
            return true
        } else if val<target {
            l=mid+1
        } else {
            r=mid-1
        }
    }
    return false
}

File: 0074 search-a-2d-matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int l = 0, r = m*n-1;
        while( l <= r ) {
            int mid = l+(r-l)/2;
            int ro = mid/n;
            int c = mid%n;
            if(matrix[ro][c]==target) return true;
            else if(matrix[ro][c] < target) l=mid+1;
            else r=mid-1;
        }
        return false;
    }
};
